presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -586,7 +788,35 @@ private void flattenNode(PlanNode node, int limit)
 
             MultiJoinNode toMultiJoinNode()
             {
-                return new MultiJoinNode(sources, and(filters), outputVariables);
+                ImmutableSet<VariableReferenceExpression> inputVariables = sources.stream().flatMap(source -> source.getOutputVariables().stream()).collect(toImmutableSet());
+
+                // We could have some output variables that were possibly generated from intermediate assignments
+                // For each of these variables, use the intermediate assignments to replace this variable with the set of input variables it uses
+
+                // Additionally, we build an overall set of assignments for the reordered Join node - this is used to add a wrapper Project over the updated output variables
+                // We do this to satisfy the invariant that the rewritten Join node must produce the same output variables as the input Join node
+                ImmutableSet.Builder<VariableReferenceExpression> updatedOutputVariables = ImmutableSet.builder();
+                Assignments.Builder overallAssignments = Assignments.builder();
+                boolean nonIdentityAssignmentsFound = false;
+
+                for (VariableReferenceExpression outputVariable : outputVariables) {
+                    if (inputVariables.contains(outputVariable)) {
+                        overallAssignments.put(outputVariable, outputVariable);
+                        updatedOutputVariables.add(outputVariable);
+                        continue;
+                    }
+
+                    checkState(intermediateAssignments.getMap().containsKey(outputVariable),
+                            "Output variable [%s] not found in input variables or in intermediate assignments", outputVariable);
+                    nonIdentityAssignmentsFound = true;
+                    overallAssignments.put(outputVariable, intermediateAssignments.get(outputVariable));
+                    updatedOutputVariables.addAll(extractUnique(intermediateAssignments.get(outputVariable)));
```

Comment:
Have you tested with the case where multiple assignments in the join nodes, and one assignment use the output of the another assignments as input? For example, one assignment is a := b+c, while c := d+e in another assignment. Here only b, d, e are from source input, and this will likely to fail again.